File: csharp/Structural/12-Proxy/Proxy/LazyJobCardRegistry.cs
public class LazyJobCardRegistry : IJobCardAccess
{
    private IJobCardAccess? _realRegistry;

    private IJobCardAccess Registry =>
        _realRegistry ??= CreateRegistry();

    private IJobCardAccess CreateRegistry()
    {
        Console.WriteLine("  [LazyProxy] Initialising real registry (first access)...");
        return new JobCardRegistry();
    }

    public void Register(JobCard card) => Registry.Register(card);
    public JobCard? GetJobCard(string jobId) => Registry.GetJobCard(jobId);
    public IReadOnlyList<JobCard> GetAll() => Registry.GetAll();
}

File: typescript/Structural/12-Proxy/Proxy/LazyJobCardRegistry.ts
import { IJobCardAccess } from "../Subject/IJobCardAccess";
import { JobCard } from "../Model/JobCard";
import { JobCardRegistry } from "../Real/JobCardRegistry";

export class LazyJobCardRegistry implements IJobCardAccess {
  private realRegistry?: JobCardRegistry;

  private get registry(): JobCardRegistry {
    if (!this.realRegistry) {
      console.log("  [LazyProxy] Initialising real registry (first access)...");
      this.realRegistry = new JobCardRegistry();
    }
    return this.realRegistry;
  }

  register(card: JobCard): void {
    this.registry.register(card);
  }

  getJobCard(jobId: string): JobCard | undefined {
    return this.registry.getJobCard(jobId);
  }

  getAll(): ReadonlyArray<JobCard> {
    return this.registry.getAll();
  }
}
